runtime/Java/src/org/antlr/v4/runtime/ParserInterpreter.java

```diff
@@ -404,4 +404,46 @@ public void addDecisionOverride(int decision, int tokenIndex, int forcedAlt) {
 		overrideDecisionInputIndex = tokenIndex;
 		overrideDecisionAlt = forcedAlt;
 	}
+
+	/** Rely on the error handler for this parser but, if no tokens are consumed
+	 *  to recover, add an error node. Otherwise, nothing is seen in the parse
+	 *  tree.
+	 */
+	protected void recover(RecognitionException e) {
+		int i = _input.index();
+		getErrorHandler().recover(this, e);
+		if ( _input.index()==i ) {
+			// no input consumed, better add an error node
+			if ( e instanceof InputMismatchException ) {
+				InputMismatchException ime = (InputMismatchException)e;
+				Token tok = e.getOffendingToken();
+				int expectedTokenType = ime.getExpectedTokens().getMinElement(); // get any element
+				String tokenText;
+				if ( expectedTokenType== Token.EOF ) tokenText = "<missing EOF>";
+				else tokenText = "<mismatched "+tok.getText()+">";
+
+				Token errToken =
+					getTokenFactory().create(new Pair<TokenSource, CharStream>(tok.getTokenSource(), tok.getTokenSource().getInputStream()),
+				                             expectedTokenType, tok.getText(),
+				                             Token.DEFAULT_CHANNEL,
+				                            -1, -1, // invalid start/stop
```

Comment:
:bulb: If you pass the start/stop from `tok` instead of invalid, you can avoid the expensive call to `tok.getText()` without losing the ability to show the text for an error.